org.eclipse.scanning.device.ui/src/org/eclipse/scanning/device/ui/model/ModelPersistAction.java

```diff
@@ -0,0 +1,152 @@
+import org.eclipse.richbeans.widgets.file.FileSelectionDialog;
+import org.eclipse.scanning.api.IModelProvider;
+import org.eclipse.scanning.api.stashing.IStashing;
+import org.eclipse.scanning.device.ui.Activator;
+import org.eclipse.scanning.device.ui.ServiceHolder;
+import org.eclipse.swt.widgets.Display;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * 
+ * This action can be used to allow scan paths to 
+ * be saved and loaded.
+ * 
+ * @author Matthew Gerring
+ *
+ */
+public class ModelPersistAction<T> extends Action {
+	
+	private static final Logger logger = LoggerFactory.getLogger(ModelPersistAction.class);
+
+	public static final String[] IDS = new String[]{"SAVE", "LOAD"};
+	/**
+	 * The action has two types, save and load.
+	 * 
+	 * @author Matthew Gerring
+	 *
+	 */
+	public static enum PersistType {
+		
+		SAVE("Save list to file",   "icons/folder-export.png"), 
+		LOAD("load list from file", "icons/folder-import.png");
+		
+		private final String          label;
+		private final String          descriptor;
+
+		PersistType(String label, String descriptor) {
+		    this.label = label;
+		    this.descriptor = descriptor;
+	    }
+
+		public String getLabel() {
+			return label;
+		}
+
+		public String getDescriptor() {
+			return descriptor;
+		}
+	}
+
+	private final PersistType      type;
+	private final IModelProvider<T> provider;
+	private       Class<T>       modelClass;
+	
+	public ModelPersistAction(IModelProvider<T> provider, PersistType type) {
+		this(provider, type, null);
+	}
+
+	/**
+	 * 
+	 * @param type
+	 */
+	public ModelPersistAction(IModelProvider<T> provider, PersistType type, Class<T> modelClass) {
+		super(type.getLabel(), Activator.getImageDescriptor(type.getDescriptor()));
+		setId(type.toString());
+		this.provider = provider;
+		this.type     = type;
+		this.modelClass = modelClass;
+	}
+	
+	@Override
+	public void run() {
+		
+		try {
+			if (type==PersistType.LOAD) {
+				doLoad();
+			} else if (type==PersistType.SAVE) {
+				doSave();
+			}
+		} catch (Exception neio) {
+			logger.error("Cannot read directory for saved files!", neio);
+		}
+	}
+
+	private void doSave() throws Exception {
+		IStashing stash = chooseFile(type);
+		if (stash==null) return; // They cancelled it
+		stash.save(provider.getModel());
+	}
+
+	private void doLoad() throws IOException {
+		IStashing stash = chooseFile(type);
+		if (stash==null) return; // They cancelled it
+		T path = stash.load(modelClass);
+		try {
+			provider.updateModel(path); // Ensures that validate and events are fired.
+		} catch (Exception e) {
+			logger.error("Problem setting the model. It might be the wrong model type!", e);
+		}
+	}
+
+	private static String previousFile = System.getProperty("GDA/gda.var", System.getProperty("user.home"));
+	private final static String[] extensions = new String[]{"json", "*.*"};
+	private final static String[] files = new String[]{"Model Files (json)", "All Files"};

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AModifiersOrderCheck)